cirq/ops/projectors.py

```diff
@@ -0,0 +1,103 @@
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+    from cirq import protocols
+
+
+@value.value_equality
+class Projector(raw_types.Gate):
+    """A non-unitary gate that projects onto one or more qubits.
+
+    The input is a matrix representing the basis of the space we project onto.
+    The basis vectors need not be orthogonal, but they must be independent (i.e.
+    the matrix has full rank).
+
+    For example, if you want to project on |0>, you would provide the basis
+    [[1, 0]]. To project onto |10>, you would provide the basis [[0, 0, 1, 0]].
+    If you want to project on the space spanned by |10> and |11>, you could
+    provide the basis [[0, 0, 1, 0], [0, 0, 0, 1]].
+    """
+
+    def __init__(self,
+                 projection_basis: Union[List[List[float]], np.ndarray],
+                 qid_shape: Tuple[int, ...] = (2,),
+                 enforce_orthonormal_basis: bool = False):
+        """
+        Args:
+            projection_basis: a (p, 2**num_qubits) matrix that lists the
+                projection vectors, where p is the dimension of the subspace
+                we're projecting on. If you project onto a single vector, then
+                p = 1 and thus the matrix reduces to a single row.
+            qid_shape: Specifies the dimension of the projection. The default is
+                2 (qubit).
+            enforce_orthonormal_basis: Whether to enfore the input basis to be
+                orthogonal.
+
+        Raises:
+            ValueError: If the basis vector is empty.
+        """
+        projection_array = np.asarray(projection_basis)
+
+        if len(projection_array.shape) != 2:
+            raise ValueError('The input projection_basis must be a 2D array')
+
+        if enforce_orthonormal_basis:
+            B = projection_array @ projection_array.T.conj()
+            if not np.allclose(B, np.eye(projection_array.shape[0])):
+                raise ValueError('The basis must be orthogonal')
+
+        if np.linalg.matrix_rank(projection_array) < projection_array.shape[0]:
+            raise ValueError('Vectors in basis must be linearly independent')
+
+        if np.prod(qid_shape) != projection_array.shape[1]:
+            raise ValueError(
+                "Invalid shape " +
+                f"{np.array(projection_array).shape} for qid_shape {qid_shape}")
+        self._projection_basis = projection_array
+        self._qid_shape = qid_shape
+
+    def _projection_basis_(self) -> np.ndarray:
+        return self._projection_basis
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    def _has_unitary_(self) -> bool:
+        return False
+
+    def _is_parameterized_(self) -> bool:
+        return False
+
+    def _channel_(self) -> Iterable[np.ndarray]:
+        A = self._projection_basis
+        pseudoinverse = A.T.conj() @ np.linalg.inv(A @ A.T.conj())
```

Comment:
I'm confused: `A @ A.T.conj()` (again conj() seems unnecessary) is identity, isn't it? We are ensuring that in the constructor. In this case pseudoinverse is `A.T.conj()`?

Reply:
Remember that A is not a square matrix. It has 2^n columns, but only p rows where p < 2^n is the dimension of the subspace we're projecting onto.

(I got confused about this, too. Suggested to name the intermediate result "pseudoinverse" to improve readability. Other ideas welcome!)

Reply:
Oh I see now the orthonormality is not enforced always :) I'm getting there slowly.

Reply:
I'm wondering if `numpy.linalg.pinv` would be better or not?

Reply:
I'm also confused about using our basis vectors as rows. Most places I read the projector to the _range_ (i.e. column space) of the matrix A is calculated as `A @ left_inverse` - however here we calculate the `right_inverse @ A` - which seems to be the range of `A.conj().T` - which is not what we want - and might explain the difference the cirq.KET_IMAGE discrepancy as well.

Reply:
Here's my suggestion: 

1. Make basis vectors column vectors 
2. Calculate left inverse using np.linalg.pinv
3. Add a comment explaining exactly what's happening: "As A's column vectors are linearly independent, A@A.conj().T is invertible, and we can calculate the left pseudoinverse. A @ left_inverse is then the orthogonal projector to the range of A (i.e. the subspace spanned by A's column vectors)

WDYT?

Reply:
In short this is the correct solution if we don't want to change the basis vector representation altogether.

```
       # make rows into columns 
        A = self._projection_basis.T
        # left pinverse 
        pseudoinverse = np.linalg.pinv(A)
        # projector to the range of A 
        return (A @ pseudoinverse,)
```

Reply:
Done.

Also added the consistency check you had suggested below.